jack-core/src/com/rapleaf/jack/transaction/DbMetricsImpl.java

```diff
@@ -0,0 +1,159 @@
+      totalActiveTime += numActive * updateToNowTime;
+      totalConnectionWaitingTime += numWaiters * updateToNowTime;
+      totalIdleTimeMinValue += numIdle * updateToNowTime;
+      if (currentConnections > numActive) {
+        totalIdleTimeMaxValue += (currentConnections - numActive) * updateToNowTime;
+      }
+      if (isOpenConnection) {
+        totalQueries += 1;
+      }
+      int newConnections = numActive + numIdle;
+      if (newConnections > currentConnections) {
+        openedConnections += newConnections - currentConnections;
+      }
+      currentConnections = newConnections;
+
+      if (numActive == maxTotalConnections) {
+        maxActiveConnectionsTime += updateToNowTime;
+      }
+
+    } catch (Exception e) {
+      LOG.error("failed to update statistics", e);
+    }
+  }
+
+  @Override
+  public long getTotalQueries() {
+    return totalQueries;
+  }
+
+  @Override
+  public double getMaxConnectionsProportion() {
+    return (double)maxActiveConnectionsTime / (double)lifeTimeStopwatch.elapsedMillis();
+  }
+
+  @Override
+  public double getAverageConnectionWaitingTime() {
+    if (totalQueries > 0) {
+      return (double)totalConnectionWaitingTime / (double)totalQueries;
+    } else {
+      return -1;
+    }
+  }
+
+  @Override
+  public double getAverageConnectionExecutionTime() {
+    if (totalQueries > 0) {
+      return ((double)totalActiveTime / (double)totalQueries);
+    } else {
+      return -1;
+    }
+  }
+
+  @Override
+  public long getOpenedConnectionsNumber() {
+    return openedConnections;
+  }
+
+  @Override
+  public double getAverageIdleConnectionsMaxValue() {
+    return (double)totalIdleTimeMaxValue / (double)lifeTimeStopwatch.elapsedMillis();
+  }
+
+  @Override
+  public double getAverageIdleConnectionsMinValue() {
+    return (double)totalIdleTimeMinValue / (double)lifeTimeStopwatch.elapsedMillis();
+  }
+
+  @Override
+  public double getAverageActiveConnections() {
+    return (double)totalActiveTime / (double)lifeTimeStopwatch.elapsedMillis();
+  }
+
+  @Override
+  public long getMaxConnectionWaitingTime() {
+    return maxConnectionWaitingTime;
+  }
+
+  @Override
+  public long getLifeTime() {
+    return lifeTimeStopwatch.elapsedMillis();
+  }
+
+  @Override
+  public String getSummary() {
+    String summary = "";
+    summary += ("\n-----------------------TRANSACTOR METRICS-----------------------\n");
+
+    summary += String.format("\nAverage number of Idle connections is between %,.2f and %,.2f", getAverageIdleConnectionsMinValue(), getAverageIdleConnectionsMaxValue());
+    summary += String.format("\nAverage number of Active connections : %,.2f", getAverageActiveConnections());
+    summary += ("\nTotal number of queries/executions : " + getTotalQueries());
+    summary += ("\nConnections opened : " + getOpenedConnectionsNumber());
+    summary += String.format("\n Max capacity time (%%) : %,.2f %%", 100 * getMaxConnectionsProportion());
```

Comment:
There is an extra space before "max". Also the "(%%)" can be omitted, since it is already added after the number.